### PR TITLE
Client tests

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -141,12 +182,21 @@ func (c *populatingClient) Create(ctx context.Context, obj runtime.Object) error
...
 // Update implements Client
 func (c *populatingClient) Update(ctx context.Context, obj runtime.Object) error {
-	meta, client, resource, err := c.metaAndClientFor(obj)
+	meta, client, mapping, resource, err := c.metaAndClientFor(obj)
 	if err != nil {
 		return err
 	}
+	if isNamespaced(mapping) {
+		return client.Put().
+			Namespace(meta.GetNamespace()).
```
Comment (seans3) [inline pkg/client/client.go]: There is a version of this method called "NamespaceIfScoped" which would simplify this "if" block.

File: pkg/client/client.go [hunk 2]
```diff
@@ -156,12 +206,19 @@ func (c *populatingClient) Update(ctx context.Context, obj runtime.Object) error
...
 // Delete implements Client
 func (c *populatingClient) Delete(ctx context.Context, obj runtime.Object) error {
-	meta, client, resource, err := c.metaAndClientFor(obj)
+	meta, client, mapping, resource, err := c.metaAndClientFor(obj)
 	if err != nil {
 		return err
 	}
+	if isNamespaced(mapping) {
+		return client.Delete().
+			Namespace(meta.GetNamespace()).
```
Comment (seans3) [inline pkg/client/client.go]: Same comment about "NamespaceIfScoped"

File: pkg/client/client.go [hunk 3]
```diff
@@ -170,12 +227,19 @@ func (c *populatingClient) Delete(ctx context.Context, obj runtime.Object) error
...
 // Get implements Client
 func (c *populatingClient) Get(ctx context.Context, key ObjectKey, obj runtime.Object) error {
-	client, resource, err := c.clientFor(obj)
+	_, client, mapping, resource, err := c.metaAndClientFor(obj)
 	if err != nil {
 		return err
 	}
+	if isNamespaced(mapping) {
+		return client.Get().
+			Namespace(key.Namespace).
```
Comment (seans3) [inline pkg/client/client.go]: Same comment about "NamspaceIfScoped".

File: pkg/client/client.go [hunk 4]
```diff
@@ -184,16 +248,20 @@ func (c *populatingClient) Get(ctx context.Context, key ObjectKey, obj runtime.O
...
+	_, client, mapping, resource, err := c.metaAndClientFor(obj)
 	if err != nil {
 		return err
 	}
-	ns := ""
-	if opts != nil {
-		ns = opts.Namespace
+	if isNamespaced(mapping) && opts != nil && len(opts.Namespace) > 0 {
+		return client.Get().
+			Namespace(opts.Namespace).
```
Comment (seans3) [inline pkg/client/client.go]: Same comment about "NamespaceIfScoped"

File: pkg/client/interfaces.go [hunk 5]
```diff
@@ -15,8 +31,8 @@ type ObjectKey = types.NamespacedName
 
 // TODO(directxman12): is there a sane way to deal with get/delete options?
 
-// ReadInterface knows how to read and list Kubernetes objects.
-type ReadInterface interface {
+// Reader knows how to read and list Kubernetes objects.
+type Reader interface {
```
Comment (seans3) [inline pkg/client/interfaces.go]: Reader is a better interface name.

File: pkg/client/interfaces.go [hunk 6]
```diff
@@ -28,8 +44,8 @@ type ReadInterface interface {
 	List(ctx context.Context, opts *ListOptions, list runtime.Object) error
 }
 
-// WriteInterface knows how to create, delete, and update Kubernetes objects.
-type WriteInterface interface {
+// Writer knows how to create, delete, and update Kubernetes objects.
+type Writer interface {
```
Comment (seans3) [inline pkg/client/interfaces.go]: Writer is a better interface name.

File: pkg/client/client.go (hunk 1 above)
Comment (pwittrock) [reply pkg/client/client.go]: Done

File: pkg/client/client.go (hunk 2 above)
Comment (pwittrock) [reply pkg/client/client.go]: Done

File: pkg/client/client.go (hunk 3 above)
Comment (pwittrock) [reply pkg/client/client.go]: Done

File: pkg/client/client.go (hunk 4 above)
Comment (pwittrock) [reply pkg/client/client.go]: Done

File: pkg/client/interfaces.go (hunk 5 above)
Comment (pwittrock) [reply pkg/client/interfaces.go]: Done

File: pkg/client/interfaces.go (hunk 6 above)
Comment (pwittrock) [reply pkg/client/interfaces.go]: Done

**Comment by pwittrock** [comment]
PTAL

File: pkg/client/client.go [hunk 7]
```diff
@@ -40,161 +58,93 @@ func New(config *rest.Config, options Options) (Client, error) {
...
-	client, resource, err := c.clientFor(obj)
-	if err != nil {
-		return nil, nil, "", err
-	}
-	meta, err := meta.Accessor(obj)
-	if err != nil {
-		return nil, nil, "", err
-	}
-	return meta, client, resource, err
+// client is a client.Client that reads and writes directly from/to an API server.  It lazily initialized
```
Comment (seans3) [inline pkg/client/client.go]: Small grammar nit: It lazily "initializes" ...

**Comment by pwittrock** [comment]
Addressed the comments.  PTAL.

**Pull request merged**